### ISSUE TITLE
PLR0911 (too-many-return-statements) shouldn’t trigger for `match` statements

### ISSUE THREAD

**Issue by flying-sheep** [astral-sh/ruff#27998]
### Summary

I currently have this, which I think is by far the most readable way this code can be. It shouldn’t be flagged:

```py
def _parse_json[R: AdRef](  # noqa: PLR0911
    a: AdAcc[R], data: Sequence[str | int | None]
) -> R | LayerAcc[R] | MultiAcc[R] | GraphAcc[R]:
    match data:
        case ["layers", str() | None as l]:
            return a.layers[l]
        case ["layers", str() | None as l, *idx] if idx_2d := _as_idx_2d(idx):
            return a.layers[l][idx_2d]
        case ["obs" | "var" as dim, str() | None as col]:
            acc = a.obs if dim == "obs" else a.var
            return acc.index if col is None else acc[col]
        case ["obsm" | "varm" as dim, str(col)]:
            return (a.obsm if dim == "obsm" else a.varm)[col]
        case ["obsm" | "varm" as dim, str(col), int(i)]:
            return (a.obsm if dim == "obsm" else a.varm)[col][i]
        case ["obsp" | "varp" as dim, str(k)]:
            return (a.obsp if dim == "obsp" else a.varp)[k]
        case ["obsp" | "varp" as dim, str(k), *idx] if idx_2d := _as_idx_2d(idx):
            return (a.obsp if dim == "obsp" else a.varp)[k][idx_2d]
        case _:
            msg = f"Cannot parse {data!r}"
            raise ValueError(msg)
```

**Comment by ntBre** [comment]
This is very similar to the `if`-`else` [example](https://docs.astral.sh/ruff/rules/too-many-return-statements/#example) in the rule docs, so it seems like the kind of code the rule is meant to flag. This might be a case where a `noqa` is the right answer rather than trying to carve out an exception to the rule.

**Comment by flying-sheep** [comment]
The point of the example seems to be that there’s a more elegant solution in that case – just use a dict to dispatch.

My thinking was that a `match` is both powerful *and* has more clarity than a if-else ladder, so people often use it deliberately to untangle a large number of combinatoric cases.

**Comment by tjkuson** [comment]
Apologies if this is off-topic, but would it perhaps make sense to count control flow statements that terminate on all arms as a single statement?

E.g.,

```python
match foo:
    case bar:
        return 1
    case baz:
        return 2
```

would count as one return whereas

```python
match foo:
    case bar:
        return 1
    case baz:
        pass
    return 2
```

would count as two? The intuition is that a statement which returns on all arms is conceptually a single return (I agree with the spirit of the issue).

**Comment by ntBre** [comment]
I'll relabel the issue for consideration because I'm open to other opinions, but the rule is currently quite simple, and I'm not sure it's worth adding special cases, especially potentially tricky cases involving control flow.

https://github.com/astral-sh/ruff/blob/67dfcf04adee0a946f408fb9f4dd14f59a488e8e/crates/ruff_linter/src/rules/pylint/rules/too_many_return_statements.rs#L72-L96
